File: class5/exercises/pathSumII.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param: root: the root of binary tree
    @param: target: An integer
    @return: all valid paths
    """
    def binaryTreePathSum2(self, root, target):
        # write your code here
        result = []
        
        
        if root is None:
            return result
            
        path = [] 
        
        self.dfs(root, path, result, 0, target)
        return result
            
        
        
    def dfs(self, root, path, result, l, target):
        
        if root is None:
            return 
        
        path.append(root.val)
        
        tmp = target
        
        for i in range(l, -1, -1):
            
            tmp -= path[i]
            
            if tmp == 0:
                result.append(path[i:])
                
        self.dfs(root.left, path, result, l + 1, target)
        self.dfs(root.right, path, result, l + 1, target)
        
        path.pop()

File: interviews/Twitch/JavaTwitch/PathSumII.java
public class PathSumII {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        
        List<List<Integer>> results = new ArrayList<List<Integer>>(); 
        
        List<Integer> path = new ArrayList<Integer>();
        
        pathSumHelper(root, sum, path, results);
        
        return results;
        
    }
    
    private void pathSumHelper(TreeNode root, int sum, List<Integer> path, List<List<Integer>> results) {
        
        if (root == null) return; 
        
        if (root.left == null && root.right == null) {
            
            if (sum - root.val == 0) {
                path.add(root.val);
                results.add(new ArrayList<Integer>(path));
                path.remove(path.size() - 1);
            }
            
            return;
        }
        
        path.add(root.val);
        
        pathSumHelper(root.left, sum - root.val, path, results);
        pathSumHelper(root.right, sum - root.val, path, results);
        
        path.remove(path.size() - 1);
        
    }
}